### ISSUE TITLE
Usability - Solution to use the library as a widget 

### ISSUE THREAD

**Issue by app-generator** [app-generator/django-dynamic-datatb#3]
Hello @Mortrest 

The current version of the library works as a separate page with its own links and actions. 

We need to find a solution to provide the same features as a widget where the user can embed this dataTable element in any page. 

The **NEW** requirements:

- Handle any model enabled by `Config` (as it is now)
- Datatables View can be embedded in any page
- All Exports available  

Here are some other packages that look more or less similar: 

- https://pypi.org/search/?q=django+datatables (PyPi Search)
- https://pypi.org/project/django-serverside-datatable/
- https://pypi.org/project/django-datatables-serverside/
- https://pypi.org/project/djangorestframework-datatables/

Let's brain storm around these new requirements and let's go. 

ty!

**Comment by Mortrest** [comment]
Hi
I have an idea but may be not great.
Now, our problem is with rendering HTML of data table and inject it to another HTML pages.
We can suppose our datatb as an module that get all required fields in constructors and setters.
This object has and `.render()` function that after embedding parameters in html, return a renderable html string.
Client can pass this output to context of his/her template.
For example:
```python
from app_generator.datatb import DataTB


class MyView(views.View):
   ...
   def get(self, request):
      ...
      datatb = DataTB(
          title = "sth"
          headings = [...]
          ...
      ).render()
      ...
      return render(
         request, 'some.html', context={
                datatb=datatb
                ...
         }
      )
```
and In `some.html` file:
```html
<div>
{{datatb}}
</div>
```
I think Create, Update and Delete endpoints has no needs to change. because embedded datatb requets to these APIs.

**Comment by app-generator** [comment]
Looks clean. 
Some facts we need to take care of:

- The feature needs to target lazy devs and beginners
- We need to keep the current flexibility
  - Minimal Input from user: Model, DB Migration, and Settings Update 
- User is not forced to provide header information in view   
- If the user updates the model, no change should be required in the view
- The plugin is CSS agnostic
  - can be used with Bootstrap or Tailwind
- The new page should be rendered with an acceptable layout on mobile. 

Regarding the coding pattern, might be a good idea to take a look at the existing plugins. We might find a nice pattern. 

- https://pypi.org/project/django-serverside-datatable/
- https://pypi.org/project/django-datatables-serverside/
- https://pypi.org/project/djangorestframework-datatables/

Ty guys.

**Comment by Mortrest** [comment]
We map endpoints to database models so we recognize which model should be viewed by URL. 
I can't imagine how this data table embed in another page.
For example If We are in `localhost:8000/index.html` and have a frame that data table is rendered in it, which model selects to view?
Can you give me an example to show data table is still dynamic and embedded in another page?

**Comment by app-generator** [comment]
The page where the code runs is irrelevant and we cannot rely on the URI. 
The library should provide an object where the context is injected. 

Your initial snippet is quite closer to this pattern. 

```python
from app_generator.datatb import DataTB 

class MyView(views.View):
   ...
   def get(self, request):
      ...
      # here the user provides the MODEL 
      datatb = DataTB(
          title = "sth"
          dataSrc="Products"    # <-- Model here
      )
      ...
      return render(
         request, 'some.html', context={
                datatb=datatb.render()         # <-- the MAGIC is provided by "render()" helper
                ...
         }
      )
```

Note that we might have multiple DataTB controls on the same page.

P.S. definitely not a boring task, right? 
🚀🚀

**Comment by Mortrest** [comment]
Nice.
Your idea sounds great.
Do you think I should implement it this way?

**Comment by app-generator** [comment]
Yes @Mortrest 

Let's code it this way. 

Ty guys!

**Comment by app-generator** [comment]
P.S. Just a small additional information 

The syntax for the model managed by the widget should provide a full import PATH:

```python
from app_generator.datatb import DataTB 

class MyView(views.View):
   ...
   def get(self, request):
      ...
      # here the user provides the MODEL 
      datatb = DataTB(
          title = "sth"
          dataSrc="app1.models.Book"    # <-- FULL Path here
      )
      ...
```

The model being provided by the developer, the check in the CFG becomes obsolete (no more check if the MODEL was activated in the settings):

```python
DYNAMIC_DATATB = {
    # SLUG -> Import_PATH 
    'books'  : "app1.models.Book",
}
```

Ty Guys!

**Comment by Mortrest** [comment]
Hi
Using multiple data tables in one page needs many efforts. Because everything is dynamic. Should we make these efforts?

**Comment by app-generator** [comment]
Hello @Mortrest 
let's have a stable version for a single widget first. 
We can postpone this use-case being complex and warn the users to use a single widget/page. 

P.S. I know that this task is a hell of a job. Moving in small steps is wiser. 

Ty for the efforts.

**Comment by app-generator** [comment]
Hello @Mortrest, @mchoopani 
all good with the task?

**Comment by Mortrest** [comment]
Hello
Yes. We prepare a single widget version of this project asap.

**Comment by app-generator** [comment]
🚀🚀

**Comment by Mortrest** [comment]
Hi
Please review the pull request and give your comments.
Thanks.

**Comment by app-generator** [comment]
Hello @Mortrest @mchoopani 

Please configure this sample project for tests.

https://github.com/app-generator/django-dynamic-datatb-tests/invitations

Ty!

**Comment by app-generator** [comment]
Hello @mchoopani 

Trying to use ... some imports are missing.
The latest commits were merged to HEAD and released. 

Please fully configure this sample project to showcase the feature: 

https://github.com/app-generator/django-dynamic-services

Ty!

**Comment by Mortrest** [comment]
OK
I see some bugs while preparing the test project. We will fix the bugs, then inform you to run the project.
Thank You

**Comment by app-generator** [comment]
Ty @Mortrest 
🚀🚀

**Comment by Mortrest** [comment]
You can see using of DataTB module in [this Pull Request](https://github.com/app-generator/django-dynamic-services/pull/2).

**Comment by app-generator** [comment]
Guys ... the feature works. 
Ty for all the work. is impressive. 

A single issue remains. 
The widget is not rendered to fill the entire available space. 

The latest version of the sample showcases this problem. 
Things that need to be done to have a `production-ready` library

- Layout on AdminLTE fixed 
- Test & FIX Layout using other BS5 kits (2 or three)
- Support Tailwind and let the user choose the layout styling (via config, with defaults on BS)

All the new tasks are budgeted separately. For now, let's fix the UI for AdminLTE.

Ty! @Mortrest @mchoopani @

![image](https://user-images.githubusercontent.com/51070104/222883381-7b95e2e4-2b42-469a-9401-95d7c51f27c1.png)

**Comment by app-generator** [comment]
Hello Team, 

Some issues are fixed, but the library is far from production-ready. Here are the issues: 

- The layout is styled in different places and the code looks dirty and hard to follow 
- The old routing that uses the whole page is not working anymore
  - This feature needs to be preserved 
- The UI needs refactoring using [flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) in all components
  - Like this, the UI remains consistent in all containers.  

Let me know if the team is able to deliver the above items. 

Ty!

**Comment by app-generator** [comment]
P.S. The codebase was updated to patch somehow the UI. 
Please pull the latest changes.

**Comment by app-generator** [comment]
Closing
